packages/web-components/fast-ssr/src/element-renderer/fast-element-renderer.ts

```diff
@@ -106,37 +109,104 @@ export abstract class FASTElementRenderer extends ElementRenderer {
             );
         }
     }
+}
 
-    /**
-     * Renders the component internals to light DOM instead of shadow DOM.
-     * @param renderInfo - information about the current rendering context.
-     */
-    public *renderLight(renderInfo: RenderInfo): IterableIterator<string> {
-        // TODO - this will yield out the element's template using the template renderer, skipping any shadow-DOM specific emission.
-        yield "";
+export abstract class SyncFASTElementRenderer extends FASTElementRenderer
+    implements ElementRenderer {
+    renderAttributes = renderAttributesSync;
+    renderShadow = renderShadow;
+}
+export abstract class AsyncFASTElementRenderer extends FASTElementRenderer
+    implements AsyncElementRenderer {
+    constructor(tagName: string, renderInfo: RenderInfo) {
+        super(tagName, renderInfo);
+
+        this.element.addEventListener(PendingTaskEvent.type, this.pendingTaskHandler);
+    }
+    public *renderAttributes(
+        this: AsyncFASTElementRenderer
+    ): IterableIterator<string | Promise<string>> {
+        if (this.element !== undefined) {
+            if (this.awaiting.size) {
+                yield this.pauseRendering().then(() => "");
+            }
+            const { attributes } = this.element;
+
+            for (
+                let i = 0, name, value;
+                i < attributes.length && ({ name, value } = attributes[i]);
+                i++
+            ) {
+                if (value === "" || value === undefined || value === null) {
+                    yield ` ${name}`;
+                } else {
+                    yield ` ${name}="${escapeHtml(value)}"`;
+                }
+            }
+        }
+    }
+    renderShadow = renderShadow as (
+        renderInfo: RenderInfo
+    ) => IterableIterator<string | Promise<string>>;
+
+    private async pauseRendering() {
+        for (const awaiting of this.awaiting) {
+            try {
+                await awaiting;
+            } catch (e) {
+                // Await will throw if the Promise is rejected. In that case,
```

Comment:
I wonder if we should have any sort of hook into this so that the error can be reported/captured? Could we add the error to the element renderer's list of errors or something like that?

Reply:
Yeah, I wasn't 100% sure what to do with errors. Authors don't really have access to the ElementRenderer instance, so I don't know if pushing it there would be helpful. Maybe to the TemplateRenderer, but they aren't guarneteed to have access to that either, so maybe the RenderInfo? The promise is provided to the renderer, so presumably if the author cared about rejections, they would provide a promise with a .catch? It does seem wrong to fail silently but throwing also didn't seem right.